src/pages/DataPlaneAuthReq.tsx

```diff
@@ -1,114 +1,125 @@
-                const origUrlHostname = new URL(originalUrl).hostname;
-                if (!origUrlHostname.endsWith(`.${gatewayHost}`)) {
-                    error = 'invalid `orig_url` parameter has invalid hostname';
-                }
-            } catch (e: unknown) {
-                console.error('invalid orig_url url parameter', e);
-                error =
-                    'invalid `orig_url` parameter cannot be parsed as a URL';
-            }
-
-            if (error === null) {
-                const newUrl = new URL('/auth-redirect', originalUrl);
-                newUrl.searchParams.append('token', gatewayAuthToken);
-                newUrl.searchParams.append('orig_url', originalUrl);
-
-                logRocketConsole('redirecting after successful auth', {
-                    newUrl,
-                    gatewayUrl,
-                    originalUrl,
-                });
-                window.location.replace(newUrl.toString());
-            }
-        }
+            setRedirectResult({ error: gatewayAuthError.toString() });
+        } else if (reactorUrl && reactorToken) {
+            getLiveSpecs_dataPlaneAuthReq(catalogPrefix).then(
+                (response) => {
+                    const hostnameLabelValue =
+                        response.data?.[0]?.shard_labels.find(
+                            (label) => label.name === SHARD_LABELS.HOSTNAME
+                        )?.value;
+
+                    if (response.error || !hostnameLabelValue) {
+                        setRedirectResult({
+                            error:
+                                response.error ??
+                                `Failed to validate 'orig_url' parameter.`,
+                        });
+
+                        return;
+                    }
+
+                    const gatewayHost = reactorUrl.hostname;
+
+                    const origUrlHostname = getURL(originalUrl)?.hostname;
+
+                    if (!origUrlHostname) {
+                        setRedirectResult({
+                            error: 'invalid `orig_url` parameter cannot be parsed as a URL',
+                        });
+
+                        return;
+                    }
+
+                    // Validate that the hostname in the orig_url is a subdomain of the gateway_url and
+                    // that its subdomain matches that of the task subdomain. This is necessary in order
+                    // to prevent malicious links using an `orig_url` parameter that sends
+                    // a user's auth token to a 3rd party.
+                    if (
+                        !origUrlHostname.startsWith(hostnameLabelValue) ||
+                        !origUrlHostname.endsWith(`.${gatewayHost}`)
+                    ) {
+                        setRedirectResult({
+                            error: 'invalid `orig_url` parameter has invalid hostname',
+                        });
```

Comment:
Future Work : We should probably wrap a lot of this into a helper function that handles returning errors or the data we want.

Reply:
This was one of my largest takeaways as well.